include/sympack/cuBLAS.hpp

```diff
@@ -0,0 +1,627 @@
+                           cublasOperation_t, cublasOperation_t,
+                           Int, Int, Int,
+                           const cuDoubleComplex           *,
+                           const cuDoubleComplex           *, Int,
+                           const cuDoubleComplex           *, Int,
+                           const cuDoubleComplex           *,
+                           cuDoubleComplex           *, Int);
+
+    cublasStatus_t cublas_gemm(cublasHandle_t,
+                           cublasOperation_t, cublasOperation_t,
+                           Int, Int, Int,
+                           const __half           *,
+                           const __half           *, Int,
+                           const __half           *, Int,
+                           const __half           *,
+                           __half           *, Int);
+
+    template <typename T>
+    cublasStatus_t cublas_gemm_wrapper(
+                           cublasOperation_t opA, cublasOperation_t opB,
+                           Int M, Int N, Int K,
+                           const T           alpha,
+                           const T           * A , Int lda,
+                           const T          * B, Int ldb,
+                           const T           beta,
+                           T           * C, Int ldc) {
+        int rank;
+        int gpu_id;
+        int n_gpus;
+
+        auto start = std::chrono::system_clock::now();
+
+        MPI_Comm_rank(symPACK::world_comm, &rank);
+        cudaGetDeviceCount(&n_gpus);
+
+        gpu_id = rank % n_gpus;
+
+        T * d_A;
+        T * d_B;
+        T * d_C;
+
+        cudaSetDevice(gpu_id);
+        
+        if (opA==CUBLAS_OP_N) {
+            //A is lda*K
+            CUDA_ERROR_CHECK(cudaMalloc(reinterpret_cast<void**>(&d_A), lda * K * sizeof(A[0])));
+            CUDA_ERROR_CHECK(cudaMemcpyAsync(d_A, A, lda * K * sizeof(A[0]), cudaMemcpyHostToDevice));
+        } else if (opA==CUBLAS_OP_T) {
+            //A is lda*M
+            CUDA_ERROR_CHECK(cudaMalloc(reinterpret_cast<void**>(&d_A), lda * M * sizeof(A[0])));
+            CUDA_ERROR_CHECK(cudaMemcpyAsync(d_A, A, lda * M * sizeof(A[0]), cudaMemcpyHostToDevice));
+        }
+
+        if (opB==CUBLAS_OP_N) {
+            //B is ldb*N
+            CUDA_ERROR_CHECK(cudaMalloc(reinterpret_cast<void**>(&d_B), ldb * N * sizeof(B[0])));
+            CUDA_ERROR_CHECK(cudaMemcpyAsync(d_B, B, ldb * N * sizeof(B[0]), cudaMemcpyHostToDevice));
+        } else if (opB==CUBLAS_OP_T) {
+            //B is lda*K
+            CUDA_ERROR_CHECK(cudaMalloc(reinterpret_cast<void**>(&d_B), ldb * K * sizeof(B[0])));
+            CUDA_ERROR_CHECK(cudaMemcpyAsync(d_B, B, ldb * K * sizeof(B[0]), cudaMemcpyHostToDevice));
+        }
+        CUDA_ERROR_CHECK(cudaMalloc(reinterpret_cast<void**>(&d_C), ldc * N * sizeof(C[0])));
+        CUDA_ERROR_CHECK(cudaMemcpyAsync(d_C, C, ldc * N * sizeof(C[0]), cudaMemcpyHostToDevice));
+
+        /* Do GEMM */
```

Comment:
What is the purpose of these `cublas::cublas_FUNC_wrapper` template functions?
IIUC they only perform host/device data movement for the arguments, but don't actually perform the advertised computation! They also appear to be completed unused. This is in contrast to the `cublas::cublas_FUNC_wrapper2` template functions, that _are_ used and appear to simply invoke cuBLAS.

Is this dead code left over from some previous version? 
If they are intended to serve a useful purpose then the comment at the top of the file should probably explain, because right now their existence seems very misleading.

Reply:
You’re right, this is dead code that should not have been included in the PR. Apologies, I will remove all of this.

Reply:
Dead code has been removed in [cb5d13b](https://github.com/symPACK/symPACK/pull/12/commits/cb5d13b01d569f7077d6f51f0ea15788200b0a6c)